File: 2fa.py
#!/usr/bin/env python3
import hmac, base64, struct, hashlib, time

def totp(secret):
    secret = secret.replace(" ", "")
    secret += '=' * (-len(secret) % 8)  # Add correct '=' padding
    intervals_no = int(time.time())//30 
    key = base64.b32decode(secret, True)
    msg = struct.pack(">Q", intervals_no)
    h = hmac.new(key, msg, hashlib.sha1).digest()
    o = h[19] & 15
    d = (struct.unpack(">I", h[o:o+4])[0] & 0x7fffffff) % 1000000
    return d

import json,sys,argparse

parser = argparse.ArgumentParser()
parser.add_argument("-j", "--json", help="The 2fa json file", default="./my2fa.json")
parser.add_argument("-m", "--match", help="Only match entries that contain the provided string", default="")
parser.add_argument("-w", "--wait", help="Wait until the OTP expires and show count-down", action="store_true")
args = parser.parse_args()

try:
    myfile = open(args.json, 'rb')
    raw = myfile.read()
    myfile.close()
    data = json.loads( raw )
    for one in data['Accounts']: # Validate against the expected schema and format.
        if not isinstance(one['Name'],str) or not isinstance(one['Secret'],str) or not base64.b32decode(one['Secret']+'=' * (-len(one['Secret']) % 8),True):
            raise Exception("The json schema doesn't comply")
except Exception as err:
    print("Failed to process the json file:", err)
    sys.exit(1)

expires_at = int(time.time()+30)//30*30
for item in data['Accounts']:
    if not args.match or args.match.lower() in item['Name'].lower(): 
        print("%+12s -- %06d" % (item['Name'], totp(item['Secret'])))

print()

remaining_seconds = (expires_at - time.time())

if args.wait:
    for remaining in range(int(remaining_seconds), -1, -1):
        print("\r   remaining -- %-02d" % (remaining), end='')
        time.sleep(1)

    print()
else:
    t = time.localtime(expires_at)
    print("Expires at: %s - in %d seconds." %(time.strftime('%H:%M:%S', t), remaining_seconds))

File: 2fa.go
package main

import (
	"crypto/hmac"
	"crypto/sha1"
	"encoding/base32"
	"encoding/binary"
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"strings"
	"time"
)

type fajson struct {
	Accounts []struct {
		Name   string
		Secret string
	}
}

func main() {

	pfile := flag.String("json", "./my2fa.json", "The 2FA json file")
	pmatch := flag.String("match", "", "only match entries that contain the provided substring")
	pwait := flag.Bool("wait", false, "Wait until OTP expires and show count-down")
	flag.Parse()

	file, err := ioutil.ReadFile(*pfile)
	if err != nil {
		log.Fatal(err)
	}

	var fa fajson
	json.Unmarshal(file, &fa)

	expiresAt := (time.Now().Unix() + 30) / 30 * 30
	for _, account := range fa.Accounts {
		if len(*pmatch) == 0 || strings.Contains(strings.ToLower(account.Name), strings.ToLower(*pmatch)) {
			fmt.Printf("% 12s -- %06d\n", account.Name, totp(account.Secret))
		}
	}

	fmt.Println()

	remainingSeconds := expiresAt - time.Now().Unix()

	if *pwait {
		for remaining := remainingSeconds; remaining > -1; remaining-- {
			fmt.Printf("\r   remaining -- %-02d", remaining)
			time.Sleep(1000 * time.Millisecond)
		}

		fmt.Println()
	} else {
		fmt.Println("Expires at", time.Unix(expiresAt, 0).Format("15:04:05"), " -- in", remainingSeconds, "seconds.")
	}
}

func totp(secret string) uint32 {

	if len(secret)%8 > 0 {
		secret += strings.Repeat("=", 8-len(secret)%8)
	}

	key, err2 := base32.StdEncoding.DecodeString(strings.ToUpper(secret))
	if err2 != nil {
		log.Fatal(err2)
	}

	intervals := uint64(time.Now().Unix() / 30)

	msg := make([]byte, 8)
	binary.BigEndian.PutUint64(msg, intervals)
	mac := hmac.New(sha1.New, key)
	mac.Write(msg)
	h := mac.Sum(nil)
	o := h[19] & 15
	d := (binary.BigEndian.Uint32(h[o:o+4]) & 0x7fffffff) % 1000000
	return d
}
